predicators/approaches/grammar_search_invention_approach.py

```diff
@@ -533,34 +649,63 @@ def enumerate(self) -> Iterator[Tuple[Predicate, float]]:
                         lb2, ub2 = feature_ranges[t2][f2]
                         if abs(lb2 - ub2) < 1e-6:
                             continue
-                        # Now, we must compute the upper and lower bounds of
-                        # the expression |t1.f1 - t2.f2|. If the intervals
-                        # [lb1, ub1] and [lb2, ub2] overlap, then the lower
-                        # bound of the expression is just 0. Otherwise, if
-                        # lb2 > ub1, the lower bound is |ub1 - lb2|, and if
-                        # ub2 < lb1, the lower bound is |lb1 - ub2|.
-                        if utils.f_range_intersection(lb1, ub1, lb2, ub2):
-                            lb = 0.0
-                        else:
-                            lb = min(abs(lb2 - ub1), abs(lb1 - ub2))
-                        # The upper bound for the expression can be
-                        # computed in a similar fashion.
-                        ub = max(abs(ub2 - lb1), abs(ub1 - lb2))
-
-                        # Scale the constant by the correct range.
-                        k = constant * (ub - lb) + lb
-                        # Create classifier.
-                        comp, comp_str = le, "<="
-                        diff_classifier = _AttributeDiffCompareClassifier(
-                            0, t1, f1, 1, t2, f2, k, constant_idx, comp,
-                            comp_str)
-                        name = str(diff_classifier)
-                        types = [t1, t2]
-                        pred = Predicate(name, types, diff_classifier)
-                        assert pred.arity == 2
+                        pred = self._construct_binary_pred(
+                            t1, t2, f1, f2, constant, ub1, ub2, lb1, lb2,
+                            constant_idx)
                         yield (pred, 2 + cost
                                )  # cost = arity + cost from constant
 
+    def _enumerate_transition_changes(
```

Comment:
we can leave this for the future, but it feels like we should be able to refactor this and the other class to share some of this logic

Reply:
agreed - i couldn't see an immediately-clean way to do it tho...